Userland/Libraries/LibSoftGPU/Clipper.cpp

```diff
@@ -64,8 +69,25 @@ static constexpr GPU::Vertex clip_intersection_point(GPU::Vertex const& p1, GPU:
     return out;
 }
 
-template<Clipper::ClipPlane plane>
-FLATTEN static constexpr void clip_plane(Vector<GPU::Vertex>& input_list, Vector<GPU::Vertex>& output_list)
+static GPU::Vertex clip_intersection_point_plane(GPU::Vertex const& p1, GPU::Vertex const& p2, FloatVector4 const& plane)
+{
+    auto const x1 = plane.dot(p1.eye_coordinates);
+    auto const x2 = plane.dot(p2.eye_coordinates);
+    auto const a = x1 / (x1 - x2);
+
+    GPU::Vertex out;
+    out.position = mix(p1.position, p2.position, a);
+    out.eye_coordinates = mix(p1.eye_coordinates, p2.eye_coordinates, a);
+    out.clip_coordinates = mix(p1.clip_coordinates, p2.clip_coordinates, a);
+    out.color = mix(p1.color, p2.color, a);
+    for (size_t i = 0; i < GPU::NUM_SAMPLERS; ++i)
+        out.tex_coords[i] = mix(p1.tex_coords[i], p2.tex_coords[i], a);
+    out.normal = mix(p1.normal, p2.normal, a);
+    return out;
+}
```

Comment:
I dislike having to replicate all this code... would it be possible to have a single function for the intersection logic?

Reply:
Good point. The latest commit consolidates one level of this code, while still leaving room for the optimized frustrum code. I introduced a new enum `Clipper::Space` which denotes in which space the clipping plane is specified in. This opens up the possibility for adding other functionality later down the line.